### ISSUE TITLE
Optional QL properties in schema result in generated code that references the property incorrectly

### ISSUE THREAD

**Issue by replicant1** [apollographql/apollo-android#2541]
**Note**: I am using Apollo Gradle Plugin 2.3.1

**Example 1:** Suppose I have a type called `VisitorCardIssueInput`. That type has an **optional** property called `workerIdVisiting`. The Kotlin class generated for it by Apollo looks like below. Note that the property `workerIdVisiting` starts with a lower case `w`. The reference to it that begins with `this@VisitorCardIssueInput` calls it `WorkerIdVisiting` - with an upper case 'w'. This code won't compile.

```
data class VisitorCardIssueInput(
  ...
  val workerIdVisiting: Input<Int> = Input.absent()
) : InputType {
    if (this@VisitorCardIssueInput.WorkerIdVisiting.defined) {
  ---------------------------------^ capital W - should be lower case w    
```

**Example 2:** Suppose I have the type `VisitorInput`. That type has an **optional** property called `visitorRawId` The Kotlin class generated by Apollo for `VisitorInput` looks like below. Note that the property `visitorRawId` starts with a lower case `v`. The reference to it that begins with `this@VisitorInput` calls it `VisitorRawId` - with an upper case `v`. This code won't compile.

```
  data class VisitorInput(
   ...
  val visitorRawId: Input<Int> = Input.absent()
) : InputType {
    if (this@VisitorInput.VisitorRawId.defined) {
 -------------------------^ capital V - should be lower case v   
```
Both generated classes are incredibly close to compiling. It just requires a change in capitalisation when referring to an optional field.

**Comment by martinbonnin** [comment]
This is super weird, there are unit tests for these exact case. For an exemple: https://github.com/apollographql/apollo-android/blob/c3b57ea2ba9881bd2b679ea834d5c9ce2ce4368f/apollo-compiler/src/test/graphql/com/example/input_object_type/type/ReviewInput.kt#L109

Can you confirm that the error happens in the `marshaller` function?

```kotlin
data class VisitorCardIssueInput(
  ...
  val workerIdVisiting: Input<Int> = Input.absent()
) : InputType {
  // There should be a marshaller() function here
  override fun marshaller(): InputFieldMarshaller = InputFieldMarshaller.invoke { writer ->
    if (this@VisitorCardIssueInput.WorkerIdVisiting.defined) {
```

Also what does your `./gradlew buildEnvironment` look like?

**Comment by replicant1** [comment]
Hi Martin, 

Thanks for your quick response. One of the example files is below. The bad capitalisation is in the `marshaller()` function.

```
@Suppress("NAME_SHADOWING", "UNUSED_ANONYMOUS_PARAMETER", "LocalVariableName",
    "RemoveExplicitTypeArguments", "NestedLambdaShadowedImplicitParameter")
data class VisitorCardIssueInput(
  /**
   * The Id of the visitor who the card is being issued to.
   */
  val visitorId: Int,
  /**
   * The 25 digit code from the visitor card.
   */
  val qrCode: String,
  /**
   * The date and time the card was issued.
   */
  val date: Any,
  /**
   * The Id of the zone the visitor is signing in to.
   */
  val zoneId: Int,
  /**
   * The Id of the worker the visitor is visiting.
   */
  val workerIdVisiting: Input<Int> = Input.absent()
) : InputType {
  override fun marshaller(): InputFieldMarshaller = InputFieldMarshaller.invoke { writer ->
    writer.writeInt("VisitorId", this@VisitorCardIssueInput.visitorId)
    writer.writeString("QrCode", this@VisitorCardIssueInput.qrCode)
    writer.writeCustom("Date", CustomType.DATETIME, this@VisitorCardIssueInput.date)
    writer.writeInt("ZoneId", this@VisitorCardIssueInput.zoneId)
    if (this@VisitorCardIssueInput.workerIdVisiting.defined) {
      writer.writeInt("workerIdVisiting", this@VisitorCardIssueInput.WorkerIdVisiting.value)
    }                                       ------- bad capital -----^
  }
}
```

I have attached the output from the `./gradlew buildEnvironment here:

[buildEnvironment.txt](https://github.com/apollographql/apollo-android/files/5126912/buildEnvironment.txt)

**Comment by replicant1** [comment]
Hi Martin, 

Thanks for your quick response. More of the example file is below. 
The bad capitalisation is in the marshaller() function.

```
import com.apollographql.apollo.api.Input
import com.apollographql.apollo.api.InputType
import com.apollographql.apollo.api.internal.InputFieldMarshaller
import kotlin.Int
import kotlin.String
import kotlin.Suppress

/**
 * The visitor's details
 */
@Suppress("NAME_SHADOWING", "UNUSED_ANONYMOUS_PARAMETER", "LocalVariableName",
    "RemoveExplicitTypeArguments", "NestedLambdaShadowedImplicitParameter")
data class VisitorInput(
  /**
   * The visitor's name.
   */
  val name: String,
  /**
   * The visitor's mobile number.
   */
  val mobile: String,
  /**
   * The visitor's organisation.
   */
  val organisation: String,
  /**
   * The visitor's Id if they have completed a visitor induction. Note: if the visitor already has a
   * linked record in the system, it will not update the name, mobile, and organisation
   */
  val visitorRawId: Input<Int> = Input.absent()
) : InputType {
  override fun marshaller(): InputFieldMarshaller = InputFieldMarshaller.invoke { writer ->
    writer.writeString("Name", this@VisitorInput.name)
    writer.writeString("Mobile", this@VisitorInput.mobile)
    writer.writeString("Organisation", this@VisitorInput.organisation)
    if (this@VisitorInput.VisitorRawId.defined) { // <= Capital V for VisitorRawId should be 'v'.
      writer.writeInt("visitorRawId", this@VisitorInput.visitorRawId.value)
    }
  }
}
```

Here's the output of ./gradlew buildEnvironment:

[buildEnvironment.txt](https://github.com/apollographql/apollo-android/files/5126944/buildEnvironment.txt)

**Comment by martinbonnin** [comment]
Thanks a lot! I can reproduce with the schema extract. I'll work on a fix.

**Comment by replicant1** [comment]
Thanks Martin - much appreciated!

**Comment by LiamDeGrey** [comment]
@martinbonnin I've come across this same issue, and although the fix (#2543) resolves the compiler error, there is still a problem with the generated file.

In the example above we can see in the marshaller function that the 'visitorRawId' field is being written to a field name that starts with a lowercase letter. In my instance, this is incorrect as the field name should start with an uppercase letter - 'VisitorRawId'. Because of this invalid field name, GraphQL mutations fail with an unrecognised input field error.
`
writer.writeInt("visitorRawId", this@VisitorInput.visitorRawId.value)
`

Again, this only seems to occur on optional Int fields.

**Comment by martinbonnin** [comment]
@LiamDeGrey right, good catch! Looks like there was another place where `name` and `schemaName` were interchanged. Should be fixed with https://github.com/apollographql/apollo-android/pull/2746
